fix(transition): support transition to teleport component child

yangxiuxiu1115 opened the pull request:
fix #11910 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved transitions for components that render teleported content.
  * Transition states and animations now apply correctly to the actual rendered content.
  * Improved transition behavior for components using cached or dynamically changing content.

* **Tests**
  * Added coverage for enter and leave transitions, including final content placement for teleported elements.
  * Added regression coverage for empty teleported content and changing transition states.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
feat(transition): support transition to teleport component child

Commit:
chore: update

Commit:
chore: test

Commit:
chore: update

edison1105 commented:
/ecosystem-ci run

edison1105 commented on packages/runtime-core/src/componentRenderUtils.ts:
```diff
@@ -248,13 +248,14 @@ export function renderComponentRoot(
   // inherit transition data
   if (vnode.transition) {
-    if (__DEV__ && !isElementRoot(root)) {
+    const child = getInnerChild(root) ?? root
```
```suggestion
    const child = getInnerChild(root) || root
```
see [generated code](https://www.babeljs.cn/repl#?browsers=safari%20%3E%3D%209%2Cchrome%3E%3D23&build=&builtIns=usage&corejs=3.21&spec=false&loose=false&code_lz=MYewdgzgLgBAhjAvDAjAKFJWAjJ8YD8BqG40MweCAPtakA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Creact&prettier=false&targets=&version=7.25.6&externalPlugins=&assumptions=%7B%7D)

edison1105 commented on packages/vue/__tests__/e2e/Transition.spec.ts:
```diff
@@ -2310,7 +2310,86 @@ describe('e2e: Transition', () => {
       E2E_TIMEOUT,
-    )
+    ),
+      test(
```
There are empty lines between each test case

Commit:
chore: update

yangxiuxiu1115 committed:
Merge branch 'main' into fix/transition

yangxiuxiu1115 committed:
Merge branch 'main' into fix/transition

gcaaa31928 commented:
Another real-world case for this: we maintain a Vue 3 component library whose Portal component renders a <Teleport> as its root (it owns popup alignment + z-index). Any consumer writing the natural thing —

```
<Transition name="fade">
  <Portal v-if="open"> … </Portal>
</Transition>
```

— gets no transition at all, plus the Component inside <Transition> renders non-element root node that cannot be animated warning. Reproduced on 3.5.27.

We ended up polyfilling this PR's exact approach in userland: inside the portal's render we read getCurrentInstance().vnode.transition and call setTransitionHooks() on the vnode we place inside the <Teleport>. Both enter and leave then behave correctly — a small data point that forwarding the hooks to the teleported child is the right fix.

Is this targeted at a particular minor? We're happy to keep the workaround until then, just want to know whether to plan around it.

edison1105 committed:
Merge branch 'main' into fix/transition

edison1105 committed:
fix(transition): handle Teleport component roots correctly

Limit inner-child resolution to Teleport roots and propagate updated
transition hooks through mounted component subtrees.

edison1105 commented:
/ecosystem-ci run
